File: agent_starter_pack/base_templates/java/src/test/java/{{cookiecutter.java_package_path}}/e2e/load_test/LoadTest.java
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package {{cookiecutter.java_package}}.e2e.load_test;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

/**
 * Load tests for the A2A server using concurrent users.
 *
 * <p>Configuration via system properties or environment variables:
 * <ul>
 *   <li>staging.url or _STAGING_URL: Target server URL (required)</li>
 *   <li>load.duration: Test duration in seconds (default: 30)</li>
 *   <li>load.users: Number of concurrent users (default: 10)</li>
 *   <li>load.ramp: Ramp-up rate in users per second (default: 2.0)</li>
 *   <li>_ID_TOKEN: Bearer token for authentication (optional)</li>
 * </ul>
 *
 * <p>Run with: mvn test -Dtest=LoadTest -Dstaging.url=http://localhost:8080
 */
@Tag("load")
class LoadTest {

  private static final String A2A_ENDPOINT = "/a2a/remote/v1/message:send";

  private final AtomicLong totalRequests = new AtomicLong(0);
  private final AtomicLong successCount = new AtomicLong(0);
  private final AtomicLong failureCount = new AtomicLong(0);
  private final AtomicLong rateLimited = new AtomicLong(0);
  private final CopyOnWriteArrayList<Long> latencies = new CopyOnWriteArrayList<>();
  private final CopyOnWriteArrayList<String> errors = new CopyOnWriteArrayList<>();

  @Test
  void testLoad() throws Exception {
    // Configuration
    String baseUrl = getConfig("staging.url", "_STAGING_URL", null);
    if (baseUrl == null || baseUrl.isEmpty()) {
      System.out.println("No staging URL provided, skipping load test");
      return;
    }

    int durationSeconds = Integer.parseInt(getConfig("load.duration", "LOAD_DURATION", "30"));
    int users = Integer.parseInt(getConfig("load.users", "LOAD_USERS", "10"));
    double ramp = Double.parseDouble(getConfig("load.ramp", "LOAD_RAMP", "2.0"));
    String idToken = System.getenv("_ID_TOKEN");

    System.out.println("═══════════════════════════════════════════════════════════════");
    System.out.println("  A2A LOAD TEST CONFIGURATION");
    System.out.println("═══════════════════════════════════════════════════════════════");
    System.out.printf("  Target URL:       %s%n", baseUrl);
    System.out.printf("  A2A Endpoint:     %s%n", A2A_ENDPOINT);
    System.out.printf("  Duration:         %ds%n", durationSeconds);
    System.out.printf("  Concurrent Users: %d%n", users);
    System.out.printf("  Ramp-up Rate:     %.1f users/sec%n", ramp);
    System.out.println("═══════════════════════════════════════════════════════════════");

    Instant startTime = Instant.now();
    Duration duration = Duration.ofSeconds(durationSeconds);
    CountDownLatch doneLatch = new CountDownLatch(1);
    ExecutorService executor = Executors.newFixedThreadPool(users);

    // Schedule test end
    Executors.newSingleThreadScheduledExecutor().schedule(
        doneLatch::countDown, durationSeconds, TimeUnit.SECONDS
    );

    // Ramp up users
    long rampIntervalMs = (long) (1000.0 / ramp);
    for (int i = 0; i < users; i++) {
      final int userNum = i;
      executor.submit(() -> runUser(baseUrl, idToken, doneLatch, userNum));
      Thread.sleep(rampIntervalMs);
    }

    // Wait for test to complete
    doneLatch.await();
    executor.shutdownNow();
    executor.awaitTermination(5, TimeUnit.SECONDS);

    Duration totalDuration = Duration.between(startTime, Instant.now());
    printResults(totalDuration);

    // Fail if too many failures
    double failureRate = (double) failureCount.get() / totalRequests.get();
    if (failureRate > 0.1) {
      throw new AssertionError(
          String.format("Failure rate too high: %.2f%%", failureRate * 100)
      );
    }
  }

  private void runUser(String baseUrl, String idToken, CountDownLatch done, int userNum) {
    HttpClient client = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(10))
        .build();

    while (done.getCount() > 0) {
      try {
        sendA2AMessage(client, baseUrl, idToken);
        // Wait between requests (1-3 seconds)
        Thread.sleep(1000 + (System.nanoTime() % 2000));
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        break;
      } catch (Exception e) {
        addFailure(e.getMessage());
      }
    }
  }

  private void sendA2AMessage(HttpClient client, String baseUrl, String idToken)
      throws Exception {
    String messageId = UUID.randomUUID().toString();
    String requestBody = String.format("""
        {
            "jsonrpc": "2.0",
            "id": "%s",
            "method": "message/send",
            "params": {
                "message": {
                    "kind": "message",
                    "messageId": "%s",
                    "role": "user",
                    "parts": [{"kind": "text", "text": "Hello! Weather in New York?"}]
                }
            }
        }
        """, messageId, messageId);

    HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
        .uri(URI.create(baseUrl + A2A_ENDPOINT))
        .header("Content-Type", "application/json")
        .timeout(Duration.ofSeconds(60))
        .POST(HttpRequest.BodyPublishers.ofString(requestBody));

    if (idToken != null && !idToken.isEmpty()) {
      requestBuilder.header("Authorization", "Bearer " + idToken);
    }

    Instant start = Instant.now();
    HttpResponse<String> response = client.send(
        requestBuilder.build(),
        HttpResponse.BodyHandlers.ofString()
    );
    long latencyMs = Duration.between(start, Instant.now()).toMillis();

    totalRequests.incrementAndGet();

    if (response.statusCode() == 429) {
      rateLimited.incrementAndGet();
      addFailure("Rate limited (429)");
      return;
    }

    if (response.statusCode() != 200) {
      addFailure("HTTP " + response.statusCode() + ": " + response.body());
      return;
    }

    String body = response.body();

    // Check for JSON-RPC error
    if (body.contains("\"error\"")) {
      addFailure("JSON-RPC error: " + body);
      return;
    }

    // Success
    successCount.incrementAndGet();
    latencies.add(latencyMs);
  }

  private void addFailure(String error) {
    failureCount.incrementAndGet();
    if (errors.size() < 100) {
      errors.add(error);
    }
  }

  private void printResults(Duration totalDuration) {
    List<Long> sortedLatencies = new ArrayList<>(latencies);
    Collections.sort(sortedLatencies);

    double avgLatency = 0;
    long minLatency = 0;
    long maxLatency = 0;
    long p50 = 0;
    long p95 = 0;
    long p99 = 0;

    if (!sortedLatencies.isEmpty()) {
      long sum = sortedLatencies.stream().mapToLong(Long::longValue).sum();
      avgLatency = (double) sum / sortedLatencies.size();
      minLatency = sortedLatencies.get(0);
      maxLatency = sortedLatencies.get(sortedLatencies.size() - 1);
      p50 = percentile(sortedLatencies, 0.50);
      p95 = percentile(sortedLatencies, 0.95);
      p99 = percentile(sortedLatencies, 0.99);
    }

    double rps = totalRequests.get() / (totalDuration.toMillis() / 1000.0);
    double successRate = totalRequests.get() > 0
        ? (double) successCount.get() / totalRequests.get() * 100
        : 0;

    System.out.println();
    System.out.println("═══════════════════════════════════════════════════════════════");
    System.out.println("  A2A LOAD TEST RESULTS");
    System.out.println("═══════════════════════════════════════════════════════════════");
    System.out.printf("  Duration:       %.2fs%n", totalDuration.toMillis() / 1000.0);
    System.out.println();
    System.out.println("  REQUESTS");
    System.out.println("  ─────────────────────────────────────────────────────────────");
    System.out.printf("  Total:          %d%n", totalRequests.get());
    System.out.printf("  Successful:     %d (%.1f%%)%n", successCount.get(), successRate);
    System.out.printf("  Failed:         %d%n", failureCount.get());
    System.out.printf("  Rate Limited:   %d%n", rateLimited.get());
    System.out.printf("  Throughput:     %.2f req/sec%n", rps);
    System.out.println();
    System.out.println("  LATENCY (ms)");
    System.out.println("  ─────────────────────────────────────────────────────────────");
    System.out.printf("  Min:            %d%n", minLatency);
    System.out.printf("  Avg:            %.0f%n", avgLatency);
    System.out.printf("  Max:            %d%n", maxLatency);
    System.out.printf("  P50:            %d%n", p50);
    System.out.printf("  P95:            %d%n", p95);
    System.out.printf("  P99:            %d%n", p99);
    System.out.println("═══════════════════════════════════════════════════════════════");

    if (!errors.isEmpty()) {
      System.out.println();
      System.out.println("  ERRORS (first 5)");
      System.out.println("  ─────────────────────────────────────────────────────────────");
      for (int i = 0; i < Math.min(5, errors.size()); i++) {
        System.out.printf("  %d. %s%n", i + 1, errors.get(i));
      }
      if (errors.size() > 5) {
        System.out.printf("  ... and %d more errors%n", errors.size() - 5);
      }
    }
  }

  private long percentile(List<Long> sorted, double p) {
    if (sorted.isEmpty()) {
      return 0;
    }
    int idx = (int) ((sorted.size() - 1) * p);
    return sorted.get(idx);
  }

  private String getConfig(String sysProp, String envVar, String defaultValue) {
    String value = System.getProperty(sysProp);
    if (value == null || value.isEmpty()) {
      value = System.getenv(envVar);
    }
    if (value == null || value.isEmpty()) {
      value = defaultValue;
    }
    return value;
  }
}

File: agent_starter_pack/base_templates/go/e2e/load_test/load_test.go
//go:build load
// +build load

// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package loadtest contains load tests for the agent server using A2A protocol.
package loadtest

import (
	"bufio"
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"sort"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	"github.com/google/uuid"
)

var (
	stagingURL = flag.String("staging-url", "", "Staging server URL")
	duration   = flag.Duration("duration", 30*time.Second, "Load test duration")
	users      = flag.Int("users", 10, "Number of concurrent users")
	ramp       = flag.Float64("ramp", 0.5, "Ramp-up rate (users per second)")
)

const a2aEndpoint = "/a2a/invoke"

// LoadTestResults holds the results of the load test.
type LoadTestResults struct {
	TotalRequests int64
	SuccessCount  int64
	FailureCount  int64
	RateLimited   int64
	Latencies     []int64
	Errors        []string
	mu            sync.Mutex
	startTime     time.Time
}

func (r *LoadTestResults) addSuccess(latencyMs int64) {
	atomic.AddInt64(&r.TotalRequests, 1)
	atomic.AddInt64(&r.SuccessCount, 1)
	r.mu.Lock()
	r.Latencies = append(r.Latencies, latencyMs)
	r.mu.Unlock()
}

func (r *LoadTestResults) addFailure(err string) {
	atomic.AddInt64(&r.TotalRequests, 1)
	atomic.AddInt64(&r.FailureCount, 1)
	r.mu.Lock()
	r.Errors = append(r.Errors, err)
	r.mu.Unlock()
}

func (r *LoadTestResults) addRateLimited() {
	atomic.AddInt64(&r.RateLimited, 1)
}

func (r *LoadTestResults) percentile(p float64) int64 {
	if len(r.Latencies) == 0 {
		return 0
	}
	sorted := make([]int64, len(r.Latencies))
	copy(sorted, r.Latencies)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i] < sorted[j] })
	idx := int(float64(len(sorted)-1) * p)
	return sorted[idx]
}

// A2A JSON-RPC request structure
type jsonRPCRequest struct {
	JSONRPC string      `json:"jsonrpc"`
	Method  string      `json:"method"`
	ID      string      `json:"id"`
	Params  interface{} `json:"params"`
}

// A2A message structure
type a2aPart struct {
	Kind string `json:"kind"`
	Text string `json:"text,omitempty"`
}

type messageSendParams struct {
	Message a2aMessageWithContext `json:"message"`
}

type a2aMessageWithContext struct {
	Role      string    `json:"role"`
	Parts     []a2aPart `json:"parts"`
	ContextID string    `json:"contextId,omitempty"`
}

// TestLoad runs the load test against the staging server using A2A protocol.
func TestLoad(t *testing.T) {
	flag.Parse()

	baseURL := *stagingURL
	if baseURL == "" {
		baseURL = os.Getenv("_STAGING_URL")
	}
	if baseURL == "" {
		t.Skip("No staging URL provided, skipping load test")
	}

	idToken := os.Getenv("_ID_TOKEN")

	log.Printf("═══════════════════════════════════════════════════════════════")
	log.Printf("  A2A LOAD TEST CONFIGURATION")
	log.Printf("═══════════════════════════════════════════════════════════════")
	log.Printf("  Target URL:     %s", baseURL)
	log.Printf("  A2A Endpoint:   %s", a2aEndpoint)
	log.Printf("  Duration:       %s", *duration)
	log.Printf("  Concurrent Users: %d", *users)
	log.Printf("  Ramp-up Rate:   %.1f users/sec", *ramp)
	log.Printf("═══════════════════════════════════════════════════════════════")

	results := &LoadTestResults{startTime: time.Now()}
	var wg sync.WaitGroup
	done := make(chan struct{})

	// Start timer
	go func() {
		time.Sleep(*duration)
		close(done)
	}()

	// Ramp up users
	rampInterval := time.Duration(float64(time.Second) / *ramp)
	for i := 0; i < *users; i++ {
		wg.Add(1)
		go func(userNum int) {
			defer wg.Done()
			runUser(baseURL, idToken, results, done)
		}(i)
		time.Sleep(rampInterval)
	}

	wg.Wait()
	totalDuration := time.Since(results.startTime)

	// Calculate stats
	var avgLatency, minLatency, maxLatency float64
	var p50, p95, p99 int64
	if results.SuccessCount > 0 {
		var totalLatency int64
		minLatency = float64(results.Latencies[0])
		maxLatency = float64(results.Latencies[0])
		for _, l := range results.Latencies {
			totalLatency += l
			if float64(l) < minLatency {
				minLatency = float64(l)
			}
			if float64(l) > maxLatency {
				maxLatency = float64(l)
			}
		}
		avgLatency = float64(totalLatency) / float64(results.SuccessCount)
		p50 = results.percentile(0.50)
		p95 = results.percentile(0.95)
		p99 = results.percentile(0.99)
	}

	rps := float64(results.TotalRequests) / totalDuration.Seconds()
	successRate := float64(results.SuccessCount) / float64(results.TotalRequests) * 100

	log.Printf("")
	log.Printf("═══════════════════════════════════════════════════════════════")
	log.Printf("  A2A LOAD TEST RESULTS")
	log.Printf("═══════════════════════════════════════════════════════════════")
	log.Printf("  Duration:       %.2fs", totalDuration.Seconds())
	log.Printf("")
	log.Printf("  REQUESTS")
	log.Printf("  ─────────────────────────────────────────────────────────────")
	log.Printf("  Total:          %d", results.TotalRequests)
	log.Printf("  Successful:     %d (%.1f%%)", results.SuccessCount, successRate)
	log.Printf("  Failed:         %d", results.FailureCount)
	log.Printf("  Rate Limited:   %d", results.RateLimited)
	log.Printf("  Throughput:     %.2f req/sec", rps)
	log.Printf("")
	log.Printf("  LATENCY (ms)")
	log.Printf("  ─────────────────────────────────────────────────────────────")
	log.Printf("  Min:            %.0f", minLatency)
	log.Printf("  Avg:            %.0f", avgLatency)
	log.Printf("  Max:            %.0f", maxLatency)
	log.Printf("  P50:            %d", p50)
	log.Printf("  P95:            %d", p95)
	log.Printf("  P99:            %d", p99)
	log.Printf("═══════════════════════════════════════════════════════════════")

	// Print errors if any
	if len(results.Errors) > 0 {
		log.Printf("")
		log.Printf("  ERRORS (first 5)")
		log.Printf("  ─────────────────────────────────────────────────────────────")
		for i, err := range results.Errors {
			if i >= 5 {
				log.Printf("  ... and %d more errors", len(results.Errors)-5)
				break
			}
			log.Printf("  %d. %s", i+1, err)
		}
	}

	// Fail if too many failures
	failureRate := float64(results.FailureCount) / float64(results.TotalRequests)
	if failureRate > 0.1 {
		t.Errorf("Failure rate too high: %.2f%%", failureRate*100)
	}
}

func runUser(baseURL, idToken string, results *LoadTestResults, done <-chan struct{}) {
	client := &http.Client{Timeout: 60 * time.Second}

	for {
		select {
		case <-done:
			return
		default:
			if err := runA2AMessage(client, baseURL, idToken, results); err != nil {
				results.addFailure(err.Error())
			}
			// Wait between requests (1-3 seconds)
			time.Sleep(time.Duration(1+time.Now().UnixNano()%2) * time.Second)
		}
	}
}

func runA2AMessage(client *http.Client, baseURL, idToken string, results *LoadTestResults) error {
	headers := map[string]string{
		"Content-Type": "application/json",
	}
	if idToken != "" {
		headers["Authorization"] = "Bearer " + idToken
	}

	// Create A2A JSON-RPC request
	contextID := uuid.New().String()
	request := jsonRPCRequest{
		JSONRPC: "2.0",
		Method:  "message/send",
		ID:      uuid.New().String(),
		Params: messageSendParams{
			Message: a2aMessageWithContext{
				Role:      "user",
				Parts:     []a2aPart{{Kind: "text", Text: "Hello! Weather in New York?"}},
				ContextID: contextID,
			},
		},
	}

	startTime := time.Now()

	body, err := json.Marshal(request)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", baseURL+a2aEndpoint, bytes.NewBuffer(body))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to send A2A message: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		respBody, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("A2A failed: %d - %s", resp.StatusCode, string(respBody))
	}

	// Read full response body
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("failed to read response: %w", err)
	}

	// Try parsing as direct JSON-RPC response first
	var jsonRPCResp map[string]interface{}
	if err := json.Unmarshal(respBody, &jsonRPCResp); err == nil {
		// Check for JSON-RPC error
		if errObj, ok := jsonRPCResp["error"].(map[string]interface{}); ok {
			msg := errObj["message"]
			return fmt.Errorf("A2A error: %v", msg)
		}

		// Check for successful result with task status
		if result, ok := jsonRPCResp["result"].(map[string]interface{}); ok {
			if status, ok := result["status"].(map[string]interface{}); ok {
				if state, ok := status["state"].(string); ok {
					if state == "completed" {
						latency := time.Since(startTime).Milliseconds()
						results.addSuccess(latency)
						return nil
					}
					if state == "failed" {
						msg := status["message"]
						return fmt.Errorf("task failed: %v", msg)
					}
				}
			}
			// Result exists but no terminal state - still consider success
			latency := time.Since(startTime).Milliseconds()
			results.addSuccess(latency)
			return nil
		}
	}

	// Fall back to SSE parsing if not valid JSON-RPC
	scanner := bufio.NewScanner(bytes.NewReader(respBody))
	scanner.Buffer(make([]byte, 64*1024), 1024*1024)

	var taskCompleted bool
	var eventCount int

	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "429 Too Many Requests") {
			results.addRateLimited()
		}
		if strings.HasPrefix(line, "data: ") {
			eventJSON := strings.TrimPrefix(line, "data: ")
			var event map[string]interface{}
			if err := json.Unmarshal([]byte(eventJSON), &event); err == nil {
				eventCount++
				if result, ok := event["result"].(map[string]interface{}); ok {
					if status, ok := result["status"].(map[string]interface{}); ok {
						if state, ok := status["state"].(string); ok {
							if state == "completed" || state == "failed" || state == "canceled" {
								taskCompleted = true
							}
						}
					}
				}
			}
		}
	}

	if taskCompleted || eventCount > 0 {
		latency := time.Since(startTime).Milliseconds()
		results.addSuccess(latency)
		return nil
	}

	return fmt.Errorf("no valid A2A response received")
}
